### PR TITLE
Fix Constant numbering

### PR DESCRIPTION
**Pull request by connorjward**
Fixes https://github.com/firedrakeproject/firedrake/issues/2999.

~Requires https://github.com/FEniCS/ufl/pull/178~ (and for our fork to be updated).

I think that wider problems exist regarding our implementation of `Constant` (see [next week's meeting](https://github.com/firedrakeproject/firedrake/wiki/Firedrake-meeting-2023-07-05#cw-codegen-issues-with-the-new-constant)), but this will resolve the performance problems.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -79,6 +79,7 @@ jobs:
             --install icepack \
             --install irksome \
             --install femlium \
+            --package-branch ufl connorjward/counted-mixin2 \
```
Comment (connorjward) [inline .github/workflows/build.yml:82]:
```suggestion
            --package-branch ufl connorjward/counted-mixin \
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -79,6 +79,7 @@ jobs:
             --install icepack \
             --install irksome \
             --install femlium \
+            --package-branch ufl connorjward/test-upstream \
```
Comment (JDBetteridge) [inline .github/workflows/build.yml:82]: Drop before merge

**Review by JDBetteridge — CHANGES_REQUESTED** [review]
I think this is good to go once our UFL branch is moved forwards

File: .github/workflows/build.yml (hunk 2 above)
Comment (connorjward) [reply .github/workflows/build.yml:82]:
```suggestion
```

**Pull request merged**